Support postgresql_include in UniqueConstraint and PrimaryKeyConstraint

dlax opened the pull request:
This is supported both for schema definition and reflection.

Fixes #10665.

dlax committed:
Support postgresql_include in UniqueConstraint and PrimaryKeyConstraint

This is supported both for schema definition and reflection.

Fix #10665.

zzzeek reviewed (commented):
pretty good, just that one changelog note but we can get that whenever

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 8a6e10b1c862de4eed5ee90ffbc66409959bbf31 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 8a6e10b1c862de4eed5ee90ffbc66409959bbf31: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

dlax commented on doc/build/changelog/unreleased_20/10665.rst:
```diff
@@ -0,0 +1,11 @@
+    :tickets: 10665
+
+    Added support for ``postgresql_include`` keyword argument to
+    :class:`_schema.UniqueConstraint` and :class:`_schema.ForeignKeyConstraint`.
```
Yes indeed; should I fix this by pushing here or is this too late now this is in gerrit?

zzzeek replied:
you can push it here and i will re-pull, but in that case also set up "tags" to be

`:tags: usecase, postgresql`

dlax replied:
Ok; done.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision eaf08b8283b8575eb2c8e48cc1bb23c15d3a50ea of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset eaf08b8283b8575eb2c8e48cc1bb23c15d3a50ea added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

zzzeek commented:
something broke in one reflection test as used by the mssql dialect, and it seems to be agnostic of DBAPI.  not clear yet if this is a side effect of changes in this patch or if some other dependency has changed.  will look

zzzeek commented on lib/sqlalchemy/engine/reflection.py:
```diff
@@ -1715,6 +1715,8 @@ def _reflect_pk(
             table.primary_key.name = pk_cons.get("name")
             table.primary_key.comment = pk_cons.get("comment", None)
+            if "dialect_options" in pk_cons:
+                table.primary_key.dialect_options = pk_cons["dialect_options"]
```
this is two nested dictionaries.  you want to use dialect_kwargs and use an update, don't re-assign

```diff
diff --git a/lib/sqlalchemy/engine/reflection.py b/lib/sqlalchemy/engine/reflection.py
index 9b6835838..23d87bb7b 100644
--- a/lib/sqlalchemy/engine/reflection.py
+++ b/lib/sqlalchemy/engine/reflection.py
@@ -1715,6 +1715,10 @@ class Inspector(inspection.Inspectable["Inspector"]):
             # update pk constraint name and comment
             table.primary_key.name = pk_cons.get("name")
             table.primary_key.comment = pk_cons.get("comment", None)
+            if "dialect_options" in pk_cons:
+                table.primary_key.dialect_kwargs.update(
+                    pk_cons["dialect_options"]
+                )
 
             # tell the PKConstraint to re-initialize
             # its column collection

```

dlax replied:
Fixed this.

zzzeek reviewed (changes requested):
the change in reflection.py seems to be breaking

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -4054,6 +4117,9 @@ def _constraint_query(self, is_unique):
                     pg_catalog.pg_constraint.c.conkey, 1
                 ).label("ord"),
                 pg_catalog.pg_description.c.description,
+                pg_catalog.pg_get_constraintdef(
```
**Federico Caselli** (CaselIT) wrote:

I'm -1 on using this with a regexp. There is already a join with the index table below, I can update to use that to get this information

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

CaselIT replied:
since there's already some custom code for unique it may make sense to split unique and pk from the other constrains, since these would likely become more similar to the index reflection.

I'll work on this in the this week

dlax replied:
That's an interesting suggestion, I can try to work on this if that's okay by you.

CaselIT replied:
As you prefer, but I kinda prototyped the new query in sql yesterday, so I already have an idea at what to do.

pg_catalog shema is very messy, so it's not rhe easiest thing to work on it, sadly

dlax replied:
Added a second commit with an attempt to follow this approach. It almost works, but needs polishing (notes in the commit message)...

I'll get back to this later this week (probably on Friday); but if you want to continue this and move forwards (or propose something different), that's also fine by me.

CaselIT replied:
i already did a working version... will integrate i guess

dlax replied:
> i already did a working version... will integrate i guess

Your approach is certainly simpler; so that's fine by me. (Working on this was a good / challenging exercise for me any ways.)

CaselIT replied:
Sorry I thought you were done for the day.

I agree that the pg_catalog are challenging. I don't really know who decided they wanted them this way... You have to do innumerable joins to get anything useful out of them :(

sqla-tester commented on test/dialect/postgresql/test_compiler.py:
```diff
@@ -796,6 +797,41 @@ def test_nulls_not_distinct(self, expr_fn, expected):
+        (
+            lambda tbl: schema.AddConstraint(
+                PrimaryKeyConstraint(
+                    tbl.c.id, postgresql_include=[tbl.c.value, tbl.c.misc]
```
**Federico Caselli** (CaselIT) wrote:

let's add a column also as string

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

dlax replied:
Not sure what you mean here, `misc` is a string column. Can you elaborate?

CaselIT replied:
instead of using `tbl.c.misc` use directly the string `misc` to reference the column

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

I would prefer if we avoid regexp here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

sqla-tester commented on test/dialect/postgresql/test_compiler.py:
```diff
@@ -796,6 +797,41 @@ def test_nulls_not_distinct(self, expr_fn, expected):
+        (
+            lambda tbl: schema.AddConstraint(
+                PrimaryKeyConstraint(
+                    tbl.c.id, postgresql_include=[tbl.c.value, "misc"]
```
**Denis Laxalde** (dlax) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

sqla-tester reviewed (commented):
**Denis Laxalde** (dlax) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 1b7a3ba176779dce7afd9b97d680410fae4c1abd of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 1b7a3ba176779dce7afd9b97d680410fae4c1abd added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

dlax committed:
WIP: Retrieve "included" columns in unique/primary key constraint from the index

Replace the previous approach relying on a regex on the result of
pg_get_constraintdef() with a subquery on pg_attribute joined with
pg_index using the 'indkey' column (and excluding 'indnkeyatts').

In tests, in order to make sure the order is okay, we add more columns
to the reflected table and create the constraint with an order different
from table attributes numbering.

TODO:
- add condition on server_version for the 'include' subquery

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 1aabea7b55ece9fc0c6e069b777d4404ac01f964 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 1aabea7b55ece9fc0c6e069b777d4404ac01f964 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

sqla-tester commented on lib/sqlalchemy/engine/reflection.py:
```diff
@@ -1712,9 +1712,12 @@ def _reflect_pk(
             table.primary_key.name = pk_cons.get("name")
             table.primary_key.comment = pk_cons.get("comment", None)
+            table.primary_key.dialect_kwargs.update(
+                pk_cons.get("dialect_options", {})
```
**Michael Bayer** (zzzeek) wrote:

nit: would have been nicer to not call upon primary_key.dialect_kwargs in the commmon case that the primary key has no dialect options, as this method creates a new _DialectArgView which it memoizes and at the moment it also creates a reference cycle.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

dlax replied:
Fixed (as well as the dialect removed from test).

sqla-tester commented on test/dialect/postgresql/test_compiler.py:
```diff
@@ -796,6 +797,41 @@ def test_nulls_not_distinct(self, expr_fn, expected):
+            Column("misc", String),
+        )
+        expr = testing.resolve_lambda(expr_fn, tbl=tbl)
+        self.assert_compile(expr, expected, dialect=dd)
```
**Michael Bayer** (zzzeek) wrote:

a lot of the tests here already have `dialect=...` in the `assert_compile` but they are at least mostly unnecessary (except for those which apply special options to the dialect) since this test declares `__dialect__ = postgresql.dialect()` at the top (and even that could just read `__dialect__ = "postgresql"`, this looks like an old set of tests).

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

dlax committed:
Avoid calling primary_key.dialect_kwargs if no dialect_options

dlax committed:
Drop unnecessary dialect in compiler tests

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision e69cde4603431fe7d84fcd80346459681c7b94b7 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset e69cde4603431fe7d84fcd80346459681c7b94b7 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

CaselIT commented:
@dlax thanks for the work so far, I'll make the last changes

sqla-tester commented on lib/sqlalchemy/engine/reflection.py:
```diff
@@ -1712,9 +1712,12 @@ def _reflect_pk(
             table.primary_key.name = pk_cons.get("name")
             table.primary_key.comment = pk_cons.get("comment", None)
+            dialect_options = pk_cons.get("dialect_options")
+            if dialect_options:
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

sqla-tester commented on test/dialect/postgresql/test_compiler.py:
```diff
@@ -796,6 +797,40 @@ def test_nulls_not_distinct(self, expr_fn, expected):
+        )
+        expr = testing.resolve_lambda(expr_fn, tbl=tbl)
+        self.assert_compile(expr, expected)
+
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

should be mostly ready. will do some perf testing

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

* lib/sqlalchemy/dialects/postgresql/base.py (line 4120): Done

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

the main change is not yet ready, let's wait before doing a 2.0 change

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5818

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5818 has been **abandoned**.  That means that at least for the moment I need to close this pull request. Sorry it didn't work out :(

CaselIT commented:
> Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5818 has been **abandoned**. That means that at least for the moment I need to close this pull request. Sorry it didn't work out :(

this was a previous 2.0 backport

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

should be all done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

CaselIT commented:
thanks for getting the ball rolling on this @dlax 

if you want to chat we usually have a dev meeting on Thursdays at 17 CEST on https://app.gitter.im/#/room/#sqlalchemy_devel:gitter.im

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:



View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5818

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5818 has been **restored**.  That means I can reopen this pull request!  Hooray :)

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

looks great, very nice

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5828 has been **merged**. Congratulations! :)

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5818 has been **merged**. Congratulations! :)
